3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/GoogleAnalyticsAdapter.scala

```diff
@@ -0,0 +1,653 @@
+   */
+  private def buildCompositeContexts(
+    originalParams: Map[String, String],
+    referenceTable: List[MPData],
+    nrCompFieldsPerSchema: Map[String, Int],
+    compFieldPrefixes: List[String],
+    indicator: String
+  ): \/[String, List[(String, Map[String, FieldType])]] =
+    for {
+      // composite params have their two first chars in the list of prefixes or are cu
+      // this is done to avoid conflicts between cd custom dimension and cd screen name
+      composite  <- originalParams
+        .filterKeys(k => compFieldPrefixes.contains(k.take(2)) && (k.length > 2 || k == "cu"))
+        .right
+      brokenDown <- composite
+        .toList
+        .map { case (k, v) => breakDownCompField(k, v, indicator) }
+        .sequenceU
+      grouped     = brokenDown.flatten.groupBy(_._1).mapValues(_.map(_._2))
+      translated <-
+        grouped.foldLeft(Map.empty[String, Map[String, \/[String, Seq[FieldType]]]]) {
+          case (m, (fieldName, values)) =>
+            val additions = referenceTable
+              .filter(_.translationTable.contains(fieldName))
+              .map { d =>
+                // this is safe because of the filter above
+                val (translation, newName) = d.translationTable(fieldName)
+                val trTable = m.getOrElse(d.schemaUri, Map.empty) +
+                  (newName -> values.map(v => translation(v).disjunction).sequenceU)
+                d.schemaUri -> trTable
+              }.toMap
+            m ++ additions
+        }
+        .map { case (k, v) => (k -> v.sequenceU) }
+        .sequenceU
+      transposed  = translated.mapValues { m =>
+        val values = transpose(m.values.map(_.toList).toList)
+        values.map(m.keys zip _).map(_.toMap)
+      }
+      // we need to filter out composite contexts which might have been built unnecessarily
+      // eg due to ${indicator}pr being in 3 different schemas
+      filtered    = transposed
+        .map { case (k, vs) => k -> vs.filter(_.size > nrCompFieldsPerSchema(k)) }
+        .filter(_._2.nonEmpty)
+      flattened   = filtered.toList.flatMap { case (k, vs) => vs.map(k -> _) }
+    } yield flattened
+
+  /**
+   * Breaks down measurement protocol composite fields into a small deterministic payload.
+   * Two cases are possible:
+   *   - the composite field name ends with a value
+   *      e.g. pr12 -> val in this case the payload becomes Map(indicatorpr -> 12, pr -> val)
+   *    - the composite field name ends with a sub field name
+   *      e.g. pr12id -> val in this case the payload becomes Map(indicatorpr -> 12, prid -> val)
+   * @param fieldName raw composite field name
+   * @param value of the composite field
+   * @param indicator string used to notify that this extracted key-value pair was from the
+   * original composite field name
+   * @return a mini payload extracted from the composite field or a failure
+   */
+  private[registry] def breakDownCompField(
```

Comment:
It would probably a lot simpler and faster to use a regex for composite fields
e.g.
```
 "(pr|il|cd|cm|cg)(\\d+)([a-zA-Z]*)(\\d*)([a-zA-Z]*)(\\d*)([a-zA-Z]*)$".r
```

Reply:
mmh interesting point, I'll look into it

Reply:
went with a regex approach